### ISSUE TITLE
Support GitHub Service Hooks

### ISSUE THREAD

**Issue by throughnothing** [bigpresh/Bot-BasicBot-Pluggable-Module-GitHub#3]
It would be nice to be able to support github service hooks.  I have written this which we have been using for a while:

https://gist.github.com/1296623

If you'd be interested in getting something like this into your tree, that would be awesome.  Currently we use it to show commit updates on all of our projects, and its great because its pretty instantaneous.  I know you have the same features planned for Announce, so maybe something like this could be integrated into Announce.pm with an option to use postbacks, or poll.

**Comment by bigpresh** [comment]
Yup, that's pretty much the same approach I'm planning to integrate into Announce - a colleague has written similar code which spawns a process to listen for GitHub hook calls and notify the appropriate channels of the push.

I just wish GitHub had similar webhooks available for issue creation/updates and pull request creation/updates :)

Just out of curiosity and noseyness, can I ask where you're using B::BB::P::M::GitHub?  (If you'd rather not say, that's fine, of course!)

**Comment by throughnothing** [comment]
Yes, we have been wishing for the same service hooks for pull requests!  At the very least, I figured announce could use the service hooks for everything possible, and poll for issues/pr's.

We're using B::BB::P::M::GitHub at Crowdtilt.  We're a pretty young startup, but growing :)

**Comment by bigpresh** [comment]
"William Wolf" reply@reply.github.com wrote:

> Yes, we have been wishing for the same service hooks for pull requests!

Awesomely, GitHub have just announced just that!

I foresee some more work on this module in the very near future :-) 

## 

David Precious davidp@preshweb.co.uk
Sent from my phone so please excuse brevity / poor quoting style etc.

**Comment by throughnothing** [comment]
Wow, hadn't seen that yet.  Fantastic.  Bye bye polling :)

**Comment by bigpresh** [comment]
Even better, the hooks are configurable via the API, so the `::GitHub::Announce` module can automatically enable the appropriate hooks without the user having to go to GitHub and do it!

http://developer.github.com/v3/repos/hooks/

The only drawback is that the module would need to reliably determine the external IP/hostname of the machine it's running on in order to generate the appropriate URL to give to GitHub.  I guess that can be a configurable option, auto-determined if the option isn't set.  I'll either use whatismyip.com for that, or whack a quick script on my webserver (so I can guarantee it stays available etc).

**Comment by bigpresh** [comment]
As I mentioned in a comment on PR4, I should be able to rip out most of Announce (all the polling crap), and add some code to automatically determine the external IP/hostname of the machine we're on, pick an ephemeral port at random and fire up a listener on it, then poke GitHub's new web hooks API to arrange for web hooks to hit us with those details.

However, thinking about it, the port on which we listen will likely need to be configurable, so it can be allowed through the firewall.

I might keep it simple for now and require the user to configure a hostname/IP and port to listen for web hook triggers on, and consider automatic configuration if no config is set further down the line.

So, e.g., `!setgithubhookendpoint foo.example.com:6500`

I think I'll need to submit a patch upstream for `Net::GitHub` to add support for the new hooks API first, though.

**Comment by throughnothing** [comment]
Sounds awesome.  I've messed with the github API manually with curl and its super simple (as usual).  Look forward to seeing your changes :)

**Comment by bigpresh** [comment]
Oh, yes, I could do it directly, but since we're using Net::GitHub for the rest of the interactions, it seems sensible to try to get support into Net::GitHub ASAP and use that :)  @fayland has been pretty receptive to other (admittedly far more minor) changes, but I imagine he'll be happy to see the hooks API supported by `Net::GitHub::V3` :)

**Comment by throughnothing** [comment]
Yes, definitely.  I wasn't trying to say it shouldn't use Net::Github, I'm definitely all for that.  Just pointing out that their api's are simple, so hopefully the additions should be simple enough to Net::Github.

**Comment by fayland** [comment]
Hi guys, thanks for using Net::GitHub

do you mind to install a _dev version and test the Hook API? if it works for you, I'll bump up the version to 0.40 so that it can be configured in Makefile.PL.

http://search.cpan.org/~fayland/Net-GitHub-0.40_03/lib/Net/GitHub/V3/Repos.pm#Hooks_API

Thanks.

**Comment by fayland** [comment]
I just don't have too much time to fork the repos and submit my patch. but let me know if you have any issue with the module. I'm glad to help.

Thanks

**Comment by bigpresh** [comment]
@fayland - ah, perfect!  I didn't realise hook support was already there; I was just starting to write up hook support, as Net::GitHub::V2::Hooks - glad you pointed me at the existing support before I went much further with that :)

I will indeed test against the dev version and report back how I get on - thanks!

**Comment by bigpresh** [comment]
At first test (this is against the current GitHub code), it doesn't appear to be working:

```
re.pl$ $repos->hooks;
Use of uninitialized value $uargs[0] in sprintf at /home/davidp/dev/github/perl-net-github/lib/Net/GitHub/V3/Query.pm line 197.
"Error!"
"Not Found at (eval) line 1\n"
re.pl$ $repos->create_hook({ name => 'web', config => { url => 'http://bar.example.com/' } });
Use of uninitialized value $uargs[0] in sprintf at /home/davidp/dev/github/perl-net-github/lib/Net/GitHub/V3/Query.pm line 197.
"Error!"
"Not Found at (eval) line 1\n"
```

Testing further, but thought I'd report that first.

**Comment by fayland** [comment]
right, it totally breaks the V2 code.

usually need do

$repos->set_default_user_repo('fayland', 'perl-net-github');

before call create_hook.

or just do

$repos->create_hook($user, $repos, { name => 'web ....)

and V3 use access_token or user/pass.

Thanks.

**Comment by bigpresh** [comment]
I tried using `set_default_user_repo` and got another error - it looks like the owner passed isn't used in the API URL generated:

```
re.pl$ $repos->set_default_user_repo('bigpresh', 'Testing');
[....]
        req => bless( {
          _content => "",
          _headers => bless( {
            "accept-encoding" => "gzip",
            "content-length" => 0,
            "user-agent" => "perl-net-github 0.40"
          }, 'HTTP::Headers' ),
          _method => "GET",
          _uri => bless( do{\(my $o = "https://api.github.com/repos//Testing/hooks")}, 'URI::https' ),
          _uri_canonical => $VAR1->{ua}{page_stack}[0]{req}{_uri}
        }, 'HTTP::Request' ),
        res => bless( {
          _content => "{\"message\":\"Not Found\"}",
[...]
```

**Comment by bigpresh** [comment]
Oddly, after restarting that tinyrepl instance and running it again, the `set_default_user_repo` method gives much more sane results; using `hooks()` then returns a Not Found error:

```
re.pl$ $repos->hooks;
"Error!"
"Not Found at (eval) line 1\n"
re.pl$ $repos->hooks('bigpresh', 'Testing');
"Error!"
"Not Found at (eval) line 1\n"
```

However, going by GitHub's API docs, I should be able to get a list of configured hooks for the repo bigpresh/Testing at the URL https://api.github.com/repos/bigpresh/Testing/hooks - however, I actually get back JSON with `"message": "Not Found"` - so that's presumably where that's coming from :)

**Comment by throughnothing** [comment]
I've noticed from manual testing that with this API it appears that you _have_ to authenticate, or else I get 404's.  Could this be the issue?

**Comment by bigpresh** [comment]
Entirely possible (although if GitHub return "Not Found" when they really mean access denied, they need a beating with a clue-by-four).

**Comment by fayland** [comment]
Looks like I have lots of stuff to improve.

here is a new tar.gz fixed the content-type check issue: http://fayland.org/tmp/Net-GitHub-0.40_04.tar.gz

sample code as below:

``` perl
my $gh = Net::GitHub::V3->new( login => $ENV{GITHUB_USER}, pass => $ENV{GITHUB_PASS});
# or my $gh = Net::GitHub::V3->new( access_token => $oauth_token );
my $repos = $gh->repos;
$repos->set_default_user_repo('fayland', 'perl-net-github');
my $hook = $repos->create_hook( {
  "name" => "web",
  "active" => 'true',
  "config" => {
    "url" => "http://something.com/webhook"
  }
} );
use Data::Dumper;
diag(Dumper(\$hook));
my @hooks = $repos->hooks;
is(@hooks, 1);
my $st = $repos->delete_hook($hook->{id});
is($st, 1);
```

Authentication is changed in V3, they are not supporting the user+token anymore.

Thanks.

**Comment by throughnothing** [comment]
@bigpresh I know the old API's return 404 if you aren't atuhorized and try to view a private repo.  I think it actually makes sense so that you cant just find out what repos exist by getting unauthorized instead of 404.  Just like if you go to github.com/bigpresh/privaterepo not logged in, you'll get a 404 page in the browser even if the project exists but is private.

Of course, requiring you to be authenticated for public repos and returning a 404 if you aren't is a different issue, and it seems like it would make more sense to be unauthorized in that case.

**Comment by throughnothing** [comment]
I verified manually, and if I hit one of my public repos at:

https://api.github.com/repos/throughnothing/dotfiles/hooks

w/o auth headers, i get 404.  With auth headers, I get the proper results.  Kinda strange.

**Comment by bigpresh** [comment]
Ahhh.  Thanks @fayland - I was using `login => 'bigpresh'` and `token => '....'` ... so now I just pass a single OAuth token?  That's... annoying.

That's going to be a PITA, as OAuth seems all intended for web applications, redirecting the user etc.  I guess I'll have to register this module as an 'application' on GitHub, and add some command to perform OAuth authentication, giving the user a GItHub URL to go to or something.

**Comment by bigpresh** [comment]
Oh, fuck me, that won't even work, as they redirect back to the website in question, and their advice for "Desktop Applications" is simply "Use basic authentication for now…".

I guess we're going to have to expect users to provide their GitHub username and password to use the module; that's horrible, I don't like that idea much.

**Comment by throughnothing** [comment]
Yeah, I agree.....hmmmmmmm.  That does suck.  I wonder why they are moving away from the token....seems like a step backwards?

**Comment by fayland** [comment]
it's a known bug: https://github.com/github/developer.github.com/issues/21

they really sucks at this point. :)

Thanks

**Comment by throughnothing** [comment]
@bigpresh do you think its worth doing with just vanilla username/pass for now?  I think its a good intermediate step to at least get something usable with the new hooks, and when a better option becomes avalailable it can be updated.

**Comment by bigpresh** [comment]
Yeah, that's what I was thinking - perhaps with a note in the docs that it may be wise to create another GitHub account for your bot, and just grant the bot's account the permissions required.

**Comment by bigpresh** [comment]
Hopefully I might get a chance to work on this later on  - I made some progress, but didn't have time to finish it off.

**Comment by ywwg** [comment]
any chance this might get picked up again?
